### PR TITLE
CLI support, progress bar w/ ETA

### PR DESCRIPTION
**Pull request by stephen-riley**

### CODE REVIEW

File: Modules/ModuleFactory.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+using System.Linq;
+using Ovation.FasterQC.Net.Utils;
+
+namespace Ovation.FasterQC.Net.Modules
+{
+    public static class ModuleFactory
+    {
+        private static readonly Dictionary<string, IQcModule> moduleMap = new()
+        {
+            ["BasicStatistics"] = new BasicStatistics(),
```
Comment (michaeljon) [inline Modules/ModuleFactory.cs:11]: oooo, that's an interesting syntax. i hadn't seen that one yet

File: Modules/QualityDistributionByBase.cs [hunk 2]
```diff
@@ -7,11 +7,10 @@ public class QualityDistributionByBase : IQcModule
     {
         private const byte ILLUMINA_BASE_ADJUSTMENT = 33;
 
-        private ulong[] aQuality = new ulong[128];
-        private ulong[] cQuality = new ulong[128];
-        private ulong[] tQuality = new ulong[128];
-        private ulong[] gQuality = new ulong[128];
-
+        private readonly ulong[] aQuality = new ulong[128];
```
Comment (michaeljon) [inline Modules/QualityDistributionByBase.cs:10]: yup. thanks

File: Ovation.FasterQC.Net.csproj [hunk 3]
```diff
@@ -5,4 +5,9 @@
     <TargetFramework>net6.0</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="CommandLineParser" Version="2.8.0" />
```
Comment (michaeljon) [inline Ovation.FasterQC.Net.csproj:9]: have you tried the new one from msft yet?

File: Program.cs [hunk 4]
```diff
@@ -39,20 +49,37 @@ static void Main(string[] args)
...
-                }
+                    if (sequenceReader.SequencesRead % UpdatePeriod == 0)
+                    {
+                        Console.Error.WriteLine($"{sequenceReader.SequencesRead.WithSsiUnits()} sequences completed ({sequenceReader.ApproximateCompletion:0.0}%)");
+                    }
+                });
             }
 
-            Console.Error.WriteLine($"{sequencesProcessed} sequences processed");
+            var results = new Dictionary<string, object>()
```
Comment (michaeljon) [inline Program.cs:62]: oh cute, probably add a match to pull in the source files as additional metadata

File: Readers/BamReader.cs [hunk 5]
```diff
@@ -20,6 +18,10 @@ public class BamReader : ISequenceReader
 
         private bool disposedValue;
 
+        private int sequencesRead = 0;
+
+        public int SequencesRead => sequencesRead;
```
Comment (michaeljon) [inline Readers/BamReader.cs:23]: this is a much better place to put this

File: Readers/FastqReader.cs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 using System;
```
Comment (michaeljon) [inline Readers/FastqReader.cs:1]: i haven't profiled this vs. the line reader yet. need to do that and retire one of them. the line reader was an experiment (and the code is cleaner / safer)

File: Readers/ReaderFactory.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Ovation.FasterQC.Net.Readers
+{
+    public static class ReaderFactory
+    {
+        public static ISequenceReader Create(CliOptions settings)
+        {
+            return settings switch
+            {
+                { Fastq: true } => new FastqLineReader(settings.InputFilename, true),
```
Comment (michaeljon) [inline Readers/ReaderFactory.cs:12]: shouldn't this 'true' use the gzipped option?

File: Utils/CliOptions.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+        [Option('m', "modules", Required = true, Min = 1, HelpText = "Space-separated list of modules to run, or 'all'.")]
+        public IEnumerable<string> ModuleNames { get; set; }
+
+        public static CliOptions Settings
+        { get; set; }
+
+        public const int UpdatePeriod = 100_000;
+
+        public static void On(bool condition, Action action)
```
Comment (michaeljon) [inline Utils/CliOptions.cs:42]: hmm, though this was an extension on Bool

File: Utils/CliOptions.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            if (condition)
```
Comment (michaeljon) [inline Utils/CliOptions.cs:44]: no complaint from outside about null-check on action. weird

**Review by michaeljon — APPROVED** [review]
nice. thanks

File: Modules/ModuleFactory.cs (hunk 1 above)
Comment (stephen-riley) [reply Modules/ModuleFactory.cs:11]: yeah, that's the "new" key value pair literal syntax.

File: Ovation.FasterQC.Net.csproj (hunk 3 above)
Comment (stephen-riley) [reply Ovation.FasterQC.Net.csproj:9]: no, been using this one for a while and pretty happy with it.  🤷🏻‍♂️

File: Utils/CliOptions.cs (hunk 9 above)
Comment (stephen-riley) [reply Utils/CliOptions.cs:44]: yeah, null checks weren't turned on.  Doing that now.

File: Utils/CliOptions.cs (hunk 8 above)
Comment (stephen-riley) [reply Utils/CliOptions.cs:42]: no, just me being lazy about typing.

File: Readers/ReaderFactory.cs (hunk 7 above)
Comment (stephen-riley) [reply Readers/ReaderFactory.cs:12]: oops, yep

File: Readers/FastqReader.cs (hunk 6 above)
Comment (stephen-riley) [reply Readers/FastqReader.cs:1]: 👍🏻

File: Program.cs (hunk 4 above)
Comment (stephen-riley) [reply Program.cs:62]: added

**Pull request merged**